Simplifying LL1TkMuonFilter

trtomei opened the pull request:
#### PR description:

Addressing items 2 and 3 of https://github.com/cms-sw/cmssw/issues/39194
 - Simplified the whole applyQualities infrastructure of the module
 - Default is now `applyQualities = false` and `qualities = {}` (an empty vector). This is fine; if `applyQualities` is set to `true`, then some qualities must be furnished!

#### PR validation:

Test done:
```
cmsDriver.py Phase2 -s HLT:75e33 --processName=HLTX \
--conditions auto:phase2_realistic_T21 \
--geometry Extended2026D88 \
--era Phase2C17I13M9 \
--eventcontent FEVTDEBUGHLT \
--filein=/store/relval/CMSSW_12_5_0_pre5/RelValTTbar_14TeV/GEN-SIM-RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/10000/8abbb6aa-60d0-4d16-8a9a-7b9c95e5aee7.root \
-n 100 --nThreads 1 --no_exec
```
runs fine, and the new module interface is exercised.

trtomei committed:
Simplifying LL1TkMuonFilter

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39280/31937

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @trtomei (Thiago Tomei) for master.

It involves the following packages:

- HLTrigger/HLTfilters (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @silviodonato this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented on HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:
```diff
@@ -22,6 +22,7 @@
 #include "FWCore/Framework/interface/EventSetupRecord.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/Exception.h"
 
```
```suggestion

#include <algorithm>
```
Because of `std::binary_search`.

trtomei replied:
Thanks, done!

missirol commented on HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:
```diff
@@ -45,18 +46,23 @@ namespace {
       l1TkMuonTag_(iConfig.getParameter<edm::InputTag>("inputTag")),
-      tkMuonToken_(consumes<l1t::TrackerMuonCollection>(l1TkMuonTag_)),
-      qualityCut_(iConfig.getParameter<edm::ParameterSet>("qualities")) {
+      tkMuonToken_(consumes<l1t::TrackerMuonCollection>(l1TkMuonTag_)) {
```
```suggestion
      tkMuonToken_(consumes(l1TkMuonTag_)) {
```
Simpler (..since the PR needs to be updated anyways).

trtomei replied:
Thanks, done!

fwyzard commented on HLTrigger/HLTfilters/plugins/L1TTkMuonFilter.cc:
```diff
@@ -45,18 +46,23 @@ namespace {
   max_Eta_ = iConfig.getParameter<double>("MaxEta");
   applyQuality_ = iConfig.getParameter<bool>("applyQuality");
   applyDuplicateRemoval_ = iConfig.getParameter<bool>("applyDuplicateRemoval");
+  qualities_ = iConfig.getParameter<std::vector<int>>("qualities");
```
`qualities_` must be sorted, otherwise `std::binary_search` will not work correctly:
```c++
std::sort(qualities_.begin(), qualities_.end());
```

trtomei replied:
Thanks, done!

trtomei committed:
Update L1TTkMuonFilter.cc

 - Added `#include <algorithm>` for the `std::binary_search`
 - Simplified initialization of the `tkMuonToken_`
 - Sort the `qualities_` for the `std::binary_search`

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39280/31955

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #39280 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-405ea0/27280/summary.html
**COMMIT**: 1ad2dc08855826ac5367c8a18443ef894f12ecda
**CMSSW**: CMSSW_12_6_X_2022-09-01-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39280/27280/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3620448
 - DQMHistoTests: Total failures: 15
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3620410
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-01-1100+405ea0/52529/triggerResults): no differences found

missirol commented:
+hlt

 - addresses Items-3,4 of [#39194](https://github.com/cms-sw/cmssw/issues/39194)
 - differences in outputs of wf `39434.75` are likely due to Item-6 of [#39194](https://github.com/cms-sw/cmssw/issues/39194)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
- Backport for 12_5_X is in #39281
